iop/fs/libbdm/src/bd_cache.c

```diff
@@ -57,56 +56,71 @@ static int _read(struct block_device *bd, u32 sector, void *buffer, u16 count)
-                M_DEBUG("- CACHE HIT [block %d] [stats: read %ds, cache %ds]\n", blkidx, c->sectors_read, c->sectors_cache);
+    c->sectors_read += count;
 #endif
-                // Read from cache
-                u32 offset = (sector - c->sector[blkidx]) * 512;
-                c->lru[blkidx] = c->lru_current++;
-                memcpy(buffer, &c->cache[blkidx][offset], count * 512);
-                return count;
-            }
-        }
 
-        // Find the LRU block
-        u32 blkidx_best_lru = 0xffffffff;
-        int blkidx_best = 0;
-        for (blkidx = 0; blkidx < BLOCK_COUNT; blkidx++) {
-            if (c->lru[blkidx] < blkidx_best_lru) {
-                // Better block found
-                blkidx_best_lru = c->lru[blkidx];
-                blkidx_best = blkidx;
-            }
+    // Do a cached read
+    int blkidx;
+    for (blkidx = 0; blkidx < BLOCK_COUNT; blkidx++) {
+        if (_contains(c->sector[blkidx], sector, count)) {
+#ifdef DEBUG
+            c->sectors_cache += count;
+            //M_DEBUG("- CACHE HIT[%d] [block %d] [devread %ds, hit-ratio %d%%]\n", sector, blkidx, c->sectors_dev, (c->sectors_cache * 100) / c->sectors_read);
+#endif
+            // Minimum weight
+            if (c->weight[blkidx] < 0)
+                c->weight[blkidx] = 0;
+
+            c->weight[blkidx] += count * BLOCK_WEIGHT_FACTOR;
+
+            // Read from cache
+            u32 offset = (sector - c->sector[blkidx]) * 512;
+            memcpy(buffer, &c->cache[blkidx][offset], count * 512);
+            return count;
         }
+    }
 
+    // Find block with the lowest weight
+    int blkidx_best_weight = 0x7fffffff;
+    int blkidx_best = 0;
+    M_DEBUG("- list: ");
+    for (blkidx = 0; blkidx < BLOCK_COUNT; blkidx++) {
 #ifdef DEBUG
-        c->sectors_read  += 1; // number of reads from device
-        c->sectors_cache += 1; // number of reads from cache
-        M_DEBUG("- CACHE READ -> [block %d] [stats: read %d, cache %d]\n", blkidx_best, c->sectors_read, c->sectors_cache);
+        printf("%*d ", 3, c->weight[blkidx] / BLOCK_WEIGHT_FACTOR);
 #endif
 
-        // Fill the block
-        c->bd->read(c->bd, sector, c->cache[blkidx_best], SECTORS_PER_BLOCK);
-        c->sector[blkidx_best] = sector;
+        // Dynamic aging
+        c->weight[blkidx] -= (SECTORS_PER_BLOCK * BLOCK_WEIGHT_FACTOR / BLOCK_COUNT) + (c->weight[blkidx] / 32);
 
-        // Read from cache
-        u32 offset = (sector - c->sector[blkidx_best]) * 512;
-        c->lru[blkidx_best] = c->lru_current++;
-        memcpy(buffer, &c->cache[blkidx_best][offset], count * 512);
-        return count;
+        if (c->weight[blkidx] < blkidx_best_weight) {
+            // Better block found
+            blkidx_best_weight = c->weight[blkidx];
+            blkidx_best = blkidx;
+        }
     }
+#ifdef DEBUG
+    printf(" devread: %*d, evict %*d [%*d], add [%*d] %d\n", 4, c->sectors_dev, 2, blkidx_best, 8, c->sector[blkidx_best], 8, sector);
```

Comment:
There is an extra `%d` here without a matching argument

Reply:
Ah yes, my last minute cleanup before push. Force pushing the fix.